File: Aoc.FsSolutions/Solutions/Day05.fs
module Aoc.Solutions.Day05

open Aoc.Runner
open System.Collections.Generic
open System.Collections

type SearchState = { Min: int; Max: int }
type State = { Row: SearchState; Col: SearchState }

let midpoint s = (s.Min + s.Max) / 2

let upper (s: SearchState) = { Min = (midpoint s + 1); Max = s.Max }

let lower (s: SearchState) = { Min = s.Min; Max = midpoint s }



let search st =
    let state =
        st
        |> Seq.fold (fun state c ->
            match c with
            | 'F' ->
                { Row = lower state.Row
                  Col = state.Col }
            | 'B' ->
                { Row = upper state.Row
                  Col = state.Col }
            | 'R' ->
                { Row = state.Row
                  Col = upper state.Col }
            | 'L' ->
                { Row = state.Row
                  Col = lower state.Col }
            | _ -> state)
               { Row = { Min = 0; Max = 127 }
                 Col = { Min = 0; Max = 7 } }

    (midpoint state.Row, midpoint state.Col)

let sid (row, col) = row * 8 + col

type Day05() =
    inherit Day()

    override this.SolveA(input) =
        let seats = input.Split("\n") |> Seq.map search

        // for inp in input.Split("\n") do
        //     let seat = search inp
        //     printfn $"Seat: {inp} {seat} {id seat}"

        seats |> Seq.map sid |> Seq.max |> string

    override this.SolveB(input) =
        let seats =
            input.Split("\n") |> Seq.map search |> HashSet

        let ids =
            (Seq.map (search >> sid) (input.Split("\n")))
            |> HashSet

        let allseats =
            seq {
                for r in 0 .. 127 do
                    for c in 0 .. 7 -> (r, c)
            }

        (allseats
        |> Seq.find (fun (r, c) ->
            not (seats.Contains((r, c)))
            && ids.Contains(sid (r, c) - 1)
            && ids.Contains(sid (r, c) + 1)))
        |> sid
        |> string



    override this.Tests =
        [ Test("1", "BFFFBBFRRR", "(70, 7)", (fun input -> (search >> string) input))
          Test("2", "FFFBBBFRRR", "(14, 7)", (fun input -> (search >> string)  input))
          Test("3", "BBFFBBFRLL", "(102, 4)", (fun input -> (search >> string)  input))
          Test("4", "BFFFBBFRRR", "567", (fun input -> (search >> sid >> string)  input))
          Test("5", "FFFBBBFRRR", "119", (fun input -> (search >> sid >> string)  input))
          Test("7", "FBFBBFFRLR", "357", (fun input -> (search >> sid >> string)  input))
          Test("6", "BBFFBBFRLL", "820", (fun input -> (search >> sid >> string)  input)) ]

        |> List<Test>

File: Aoc.Solutions/Solutions/Day05.cs
using System;
using System.Collections.Generic;
using System.Linq;

using Aoc.Runner;

using Extensions;

namespace Aoc.Solutions {
    public class Day05 : Day {


        record SearchState(int Min, int Max);
        record State(SearchState Row, SearchState Col);

        static int Midpoint(SearchState st) => (st.Min + st.Max) / 2;
        static SearchState Upper(SearchState st) => st with { Min = Midpoint(st) + 1 };
        static SearchState Lower(SearchState st) => st with { Max = Midpoint(st) };
        static (int, int) Search(string seat) {
            var seatPostition =
                seat.Aggregate(new State(new(0, 127), new(0, 7)),
                    (state, c) => c switch {
                        'F' => state with { Row = Lower(state.Row) },
                        'B' => state with { Row = Upper(state.Row) },
                        'R' => state with { Col = Upper(state.Col) },
                        'L' => state with { Col = Lower(state.Col) },
                        _ => state
                    });

            return (Midpoint(seatPostition.Row), Midpoint(seatPostition.Col));
        }

        static int Sid((int, int) seat) => seat.Item1 * 8 + seat.Item2;

        public override string SolveA(string input) =>
            input.Split("\n")
                .Select(Search)
                .Select(Sid)
                .Max()
                .ToString();

        static IEnumerable<(int, int)> AllSeats() {
            foreach (var r in Enumerable.Range(0, 128))
                foreach (var c in Enumerable.Range(0, 8))
                    yield return (r, c);
        }

        public override string SolveB(string input) {
            var seats = input.Split("\n")
                .Select(Search)
                .ToHashSet();
            var ids = seats.Select(Sid).ToHashSet();
            return AllSeats().First(s =>
                !seats.Contains(s)
                && ids.Contains(Sid(s) - 1)
                && ids.Contains(Sid(s) + 1)
            ).Then(Sid).ToString();


        }
        public Day05() {
            Tests = new()
            {
                new("BFFFBBFRRR", "BFFFBBFRRR", "(70, 7)", input => Search(input).ToString())
            };
        }

    }
}